File: 2025.07/2025.07.07/1353-maximum-number-of-events-that-can-be-attended.js
import { MinPriorityQueue } from "@datastructures-js/priority-queue";

/**
 * @param {number[][]} events
 * @return {number}
 */
var maxEvents = function (events) {
  // Solution: Sort + Min Heap
  // Time: O(N * logN)
  // Space: O(N)

  let maxDay = 0;
  for (let i = 0; i < events.length; i++) {
    maxDay = Math.max(maxDay, events[i][1]);
  }

  events.sort((a, b) => a[0] - b[0]);

  const minEndDayHeap = new MinPriorityQueue();
  let count = 0;

  for (let startDay = 1, day = 0; startDay <= maxDay; startDay++) {
    while (day < events.length && events[day][0] <= startDay) {
      minEndDayHeap.enqueue(events[day][1]);
      day++;
    }

    while (minEndDayHeap.size() > 0 && minEndDayHeap.front() < startDay) {
      minEndDayHeap.dequeue();
    }

    if (minEndDayHeap.size() > 0) {
      minEndDayHeap.dequeue();
      count++;
    }

    if (day >= events.length && minEndDayHeap.size() === 0) {
      break;
    }
  }

  return count;
};

console.log(
  maxEvents([
    [1, 2],
    [2, 3],
    [3, 4],
  ])
); // 3
console.log(
  maxEvents([
    [1, 2],
    [2, 3],
    [3, 4],
    [1, 2],
  ])
); // 4
console.log(
  maxEvents([
    [1, 2],
    [1, 2],
    [3, 3],
    [1, 5],
    [1, 5],
  ])
); // 5
console.log(
  maxEvents([
    [1, 10],
    [2, 2],
    [2, 2],
    [2, 2],
    [2, 2],
  ])
); // 2

File: 2025.07/2025.07.07/1353-maximum-number-of-events-that-can-be-attended.py
import heapq
from typing import List

class Solution:
    def maxEvents(self, events: List[List[int]]) -> int:
        max_day = max(end for _, end in events)
        events.sort(key = lambda x: x[0])
        max_end_day_heap = []
        count = 0
        day = 0
        for start_day in range(1, max_day + 1):
            while day < len(events) and events[day][0] <= start_day:
                heapq.heappush(max_end_day_heap, events[day][1])
                day += 1
            while max_end_day_heap and max_end_day_heap[0] < start_day:
                heapq.heappop(max_end_day_heap)
            if max_end_day_heap:
                heapq.heappop(max_end_day_heap)
                count += 1
            if day >= len(events) and not max_end_day_heap:
                break
        return count
    
print(Solution().maxEvents([[1, 2], [2, 3], [3, 4]]))  # 3
print(Solution().maxEvents([[1, 2], [2, 3], [3, 4], [1, 2]]))  # 4
print(Solution().maxEvents([[1, 2], [1, 2], [3, 3], [1, 5], [1, 5]]))  # 5
print(Solution().maxEvents([[1, 10], [2, 2], [2, 2], [2, 2], [2, 2]]))  # 2
